src/brod_producer.erl

```diff
@@ -144,16 +144,17 @@ start_link(ClientPid, Topic, Partition, Config) ->
 produce(Pid, Key, Value) ->
   CallRef = #brod_call_ref{ caller = self()
                           , callee = Pid
-                          , ref    = make_ref()
+                          , ref    = Mref = erlang:monitor(process, Pid)
                           },
-  ok = gen_server:cast(Pid, {produce, CallRef, Key, Value}),
-  ExpectedReply = #brod_produce_reply{ call_ref = CallRef
-                                     , result   = brod_produce_req_buffered
-                                     },
-  %% Wait until the request is buffered
-  case sync_produce_request(ExpectedReply) of
-    ok              -> {ok, CallRef};
-    {error, Reason} -> {error, Reason}
+  Pid ! {produce, CallRef, Key, Value},
+  receive
```

Comment:
I've just tried, but surprisingly that also seemed to ruin the optimisation. I think the compiler just explicitly looks for variables assigned to references, so using something like 'ExpectedReply' instead of explicitly saying we want to match an Mref breaks it. I could add a comment here instead if you think that would be helpful.